# American Express Default Prediction Project
#### Mia Belli & Shruti Lakhani - BUS 316

In [2]:
import pandas as pd
import numpy as np
import sklearn.linear_model

In [17]:
data = pd.read_csv('train_data.csv', nrows = 1000000)

In [18]:
data

customer_ID         S_2  \
0       0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-03-09   
1       0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-04-07   
2       0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-05-28   
3       0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-06-13   
4       0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  2017-07-16   
...                                                   ...         ...   
999995  2e3fabd8551cfbb7819ffc99cf22d4335953e4d911b7ea...  2017-12-01   
999996  2e3fabd8551cfbb7819ffc99cf22d4335953e4d911b7ea...  2018-01-01   
999997  2e3fabd8551cfbb7819ffc99cf22d4335953e4d911b7ea...  2018-02-28   
999998  2e3fabd8551cfbb7819ffc99cf22d4335953e4d911b7ea...  2018-03-31   
999999  2e3fc76a68eeae6697b0a6b10f9288b7acaa678bbf99cd...  2017-03-29   

             P_2      D_39       B_1       B_2       R_1       S_3      D_41  \
0       0.938469  0.001733  0.008724  1.006838  0.009228  0.124035  0.008771   
1       0.936665  0.005775  0.004923  1.000653  0.006151  0.126750  0.000798   
2       0.954180  0.091505  0.021655  1.009672  0.006815  0.123977  0.007598   
3       0.960384  0.002455  0.013683  1.002700  0.001373  0.117169  0.000685   
4       0.947248  0.002483  0.015193  1.000727  0.007605  0.117325  0.004653   
...          ...       ...       ...       ...       ...       ...       ...   
999995  0.444918  0.214272  0.509806  0.024078  0.001514  0.386644  0.001295   
999996  0.445380  0.182068  0.511427  0.025369  0.008337  0.372405  0.002628   
999997  0.439738  0.155119  0.497809  0.023311  0.000641  0.389353  0.009791   
999998  0.449821  0.155417  0.501278  0.025809  0.006869  0.367082  0.001196   
999999  0.481917  0.009878  0.182513  0.192125  0.005269  0.393390  0.004422   

             B_3  ...  D_136  D_137  D_138     D_139     D_140     D_141  \
0       0.004709  ...    NaN    NaN    NaN  0.002427  0.003706  0.003818   
1       0.002714  ...    NaN    NaN    NaN  0.003954  0.003167  0.005032   
2       0.009423  ...    NaN    NaN    NaN  0.003269  0.007329  0.000427   
3       0.005531  ...    NaN    NaN    NaN  0.006117  0.004516  0.003200   
4       0.009312  ...    NaN    NaN    NaN  0.003671  0.004946  0.008889   
...          ...  ...    ...    ...    ...       ...       ...       ...   
999995  0.920215  ...    NaN    NaN    NaN  0.000700  0.004486  0.006241   
999996  0.933117  ...    NaN    NaN    NaN  0.007064  0.005506  0.005020   
999997  0.917929  ...    NaN    NaN    NaN  0.002184  0.004863  0.007474   
999998  0.926053  ...    NaN    NaN    NaN  0.000752  0.004426  0.005118   
999999  0.239441  ...    NaN    NaN    NaN  0.004562  0.003862  0.006811   

        D_142     D_143     D_144     D_145  
0         NaN  0.000569  0.000610  0.002674  
1         NaN  0.009576  0.005492  0.009217  
2         NaN  0.003429  0.006986  0.002603  
3         NaN  0.008419  0.006527  0.009600  
4         NaN  0.001670  0.008126  0.009827  
...       ...       ...       ...       ...  
999995    NaN  0.001294  0.006138  0.004346  
999996    NaN  0.006667  0.004177  0.004452  
999997    NaN  0.004654  0.005166  0.000017  
999998    NaN  0.002986  0.000797  0.006316  
999999    NaN  0.002136  0.002043  0.008634  

[1000000 rows x 190 columns]

In [19]:
# selecting variables that are only balance and deliquency variables >> because deliquency gives us a variable based off of past behavior
# while balance variables give us a current value that of what the customer owes
df=data[[ 'customer_ID','D_39', 'B_1', 'B_2', 'D_41', 'B_3', 'D_42', 'D_43', 'D_44', 'B_4', 'D_45', 'B_5', 'D_46', 'D_47', 'D_48', 'D_49', 'B_6', 'B_7', 'B_8', 'D_50', 'D_51', 'B_9', 'D_52', 'B_10', 'D_53', 'B_11', 'D_54', 'B_12', 'D_55', 'D_56', 'B_13', 'D_58', 'B_14', 'D_59', 'D_60', 'D_61', 'B_15', 'D_62', 'D_63', 'D_64', 'D_65', 'B_16', 'B_17', 'B_18', 'B_19', 'D_66', 'B_20', 'D_68', 'B_21', 'D_69', 'B_22', 'D_70', 'D_71', 'D_72', 'B_23', 'D_73', 'D_74', 'D_75', 'D_76', 'B_24', 'D_77', 'B_25', 'B_26', 'D_78', 'D_79', 'D_80', 'B_27', 'D_81', 'D_82', 'B_28', 'D_83', 'D_84', 'B_29', 'B_30', 'D_86', 'D_87', 'D_88', 'B_31', 'B_32', 'B_33', 'D_89', 'D_91', 'D_92', 'D_93', 'D_94', 'D_96', 'D_102', 'D_103', 'D_104', 'D_105', 'D_106', 'D_107', 'B_36', 'B_37', 'B_38', 'D_108', 'D_109', 'D_110', 'D_111', 'B_39', 'D_112', 'B_40', 'D_113', 'D_114', 'D_115', 'D_116', 'D_117', 'D_118', 'D_119', 'D_120', 'D_121', 'D_122', 'D_123', 'D_124', 'D_125', 'D_126', 'D_127', 'D_128', 'D_129', 'B_41', 'B_42', 'D_130', 'D_131', 'D_132', 'D_133', 'D_134', 'D_135', 'D_136', 'D_137', 'D_138', 'D_139', 'D_140', 'D_141', 'D_142', 'D_143', 'D_144', 'D_145']]

In [20]:
df

customer_ID      D_39       B_1  \
0       0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.001733  0.008724   
1       0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.005775  0.004923   
2       0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.091505  0.021655   
3       0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.002455  0.013683   
4       0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.002483  0.015193   
...                                                   ...       ...       ...   
999995  2e3fabd8551cfbb7819ffc99cf22d4335953e4d911b7ea...  0.214272  0.509806   
999996  2e3fabd8551cfbb7819ffc99cf22d4335953e4d911b7ea...  0.182068  0.511427   
999997  2e3fabd8551cfbb7819ffc99cf22d4335953e4d911b7ea...  0.155119  0.497809   
999998  2e3fabd8551cfbb7819ffc99cf22d4335953e4d911b7ea...  0.155417  0.501278   
999999  2e3fc76a68eeae6697b0a6b10f9288b7acaa678bbf99cd...  0.009878  0.182513   

             B_2      D_41       B_3  D_42      D_43      D_44       B_4  ...  \
0       1.006838  0.008771  0.004709   NaN       NaN  0.000630  0.080986  ...   
1       1.000653  0.000798  0.002714   NaN       NaN  0.002526  0.069419  ...   
2       1.009672  0.007598  0.009423   NaN       NaN  0.007605  0.068839  ...   
3       1.002700  0.000685  0.005531   NaN       NaN  0.006406  0.055630  ...   
4       1.000727  0.004653  0.009312   NaN       NaN  0.007731  0.038862  ...   
...          ...       ...       ...   ...       ...       ...       ...  ...   
999995  0.024078  0.001295  0.920215   NaN  0.385147  0.382967  0.625154  ...   
999996  0.025369  0.002628  0.933117   NaN  0.622123  0.382894  0.616366  ...   
999997  0.023311  0.009791  0.917929   NaN  0.396818  0.502761  0.645277  ...   
999998  0.025809  0.001196  0.926053   NaN  0.358734  0.503617  0.662351  ...   
999999  0.192125  0.004422  0.239441   NaN  0.097536  0.256662  0.134462  ...   

        D_136  D_137  D_138     D_139     D_140     D_141  D_142     D_143  \
0         NaN    NaN    NaN  0.002427  0.003706  0.003818    NaN  0.000569   
1         NaN    NaN    NaN  0.003954  0.003167  0.005032    NaN  0.009576   
2         NaN    NaN    NaN  0.003269  0.007329  0.000427    NaN  0.003429   
3         NaN    NaN    NaN  0.006117  0.004516  0.003200    NaN  0.008419   
4         NaN    NaN    NaN  0.003671  0.004946  0.008889    NaN  0.001670   
...       ...    ...    ...       ...       ...       ...    ...       ...   
999995    NaN    NaN    NaN  0.000700  0.004486  0.006241    NaN  0.001294   
999996    NaN    NaN    NaN  0.007064  0.005506  0.005020    NaN  0.006667   
999997    NaN    NaN    NaN  0.002184  0.004863  0.007474    NaN  0.004654   
999998    NaN    NaN    NaN  0.000752  0.004426  0.005118    NaN  0.002986   
999999    NaN    NaN    NaN  0.004562  0.003862  0.006811    NaN  0.002136   

           D_144     D_145  
0       0.000610  0.002674  
1       0.005492  0.009217  
2       0.006986  0.002603  
3       0.006527  0.009600  
4       0.008126  0.009827  
...          ...       ...  
999995  0.006138  0.004346  
999996  0.004177  0.004452  
999997  0.005166  0.000017  
999998  0.000797  0.006316  
999999  0.002043  0.008634  

[1000000 rows x 137 columns]

In [21]:
# dropping columns with null values
df1 = df.dropna(axis='columns')

In [22]:
df1

customer_ID      D_39       B_1  \
0       0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.001733  0.008724   
1       0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.005775  0.004923   
2       0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.091505  0.021655   
3       0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.002455  0.013683   
4       0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.002483  0.015193   
...                                                   ...       ...       ...   
999995  2e3fabd8551cfbb7819ffc99cf22d4335953e4d911b7ea...  0.214272  0.509806   
999996  2e3fabd8551cfbb7819ffc99cf22d4335953e4d911b7ea...  0.182068  0.511427   
999997  2e3fabd8551cfbb7819ffc99cf22d4335953e4d911b7ea...  0.155119  0.497809   
999998  2e3fabd8551cfbb7819ffc99cf22d4335953e4d911b7ea...  0.155417  0.501278   
999999  2e3fc76a68eeae6697b0a6b10f9288b7acaa678bbf99cd...  0.009878  0.182513   

             B_4       B_5      D_47       B_7      D_51       B_9      B_10  \
0       0.080986  0.170600  0.525351  0.059416  1.335856  0.008207  0.096219   
1       0.069419  0.113239  0.521311  0.057744  1.339794  0.008373  0.099804   
2       0.068839  0.060492  0.524568  0.056647  1.337179  0.009355  0.134073   
3       0.055630  0.166782  0.530929  0.049253  1.339909  0.006782  0.134437   
4       0.038862  0.143630  0.529305  0.048918  1.341735  0.000519  0.121518   
...          ...       ...       ...       ...       ...       ...       ...   
999995  0.625154  0.011167  0.230611  0.640016  0.007110  0.561398  0.014214   
999996  0.616366  0.010080  0.230187  0.615489  0.008221  0.559819  0.009806   
999997  0.645277  0.010605  0.233268  0.741261  0.001374  0.557749  0.010445   
999998  0.662351  0.015986  0.237781  0.724769  0.009544  0.556564  0.012080   
999999  0.134462  0.026001  0.464648  0.177516  0.340525  0.003448  0.040240   

        ...      B_28      D_86  B_31      B_32      D_92      D_93      D_94  \
0       ...  0.084683  0.007084     1  0.006626  1.006133  0.003569  0.008871   
1       ...  0.081843  0.006677     1  0.001854  1.005791  0.000571  0.000391   
2       ...  0.081954  0.001185     1  0.008686  1.005801  0.007425  0.009234   
3       ...  0.060634  0.003324     1  0.002478  1.007036  0.000664  0.003200   
4       ...  0.062492  0.001504     1  0.002199  1.002915  0.003079  0.003845   
...     ...       ...       ...   ...       ...       ...       ...       ...   
999995  ...  0.211481  0.004425     1  0.002744  0.009695  0.002172  0.005996   
999996  ...  0.208682  0.004128     1  0.006775  0.005867  0.004523  0.003622   
999997  ...  0.225962  0.007143     1  0.007657  0.006683  0.008812  0.005758   
999998  ...  0.226216  0.008717     1  0.005355  0.002604  0.009268  0.008626   
999999  ...  0.097136  0.004695     1  0.002233  0.007998  0.007983  0.003105   

            D_96      B_36     D_127  
0       0.004950  0.009968  1.003319  
1       0.003180  0.003921  1.008394  
2       0.005433  0.001264  1.009307  
3       0.000063  0.002729  1.001671  
4       0.000535  0.009998  1.009886  
...          ...       ...       ...  
999995  0.008168  0.004775  0.003337  
999996  0.009611  0.002064  0.000685  
999997  0.004018  0.000914  0.004696  
999998  0.006964  0.009191  0.009385  
999999  0.004076  0.008801  0.004873  

[1000000 rows x 33 columns]

In [23]:
cols = df1.columns.values.tolist()
print(cols)

['customer_ID', 'D_39', 'B_1', 'B_4', 'B_5', 'D_47', 'B_7', 'D_51', 'B_9', 'B_10', 'B_11', 'B_12', 'D_58', 'B_14', 'D_60', 'D_63', 'D_65', 'B_18', 'B_21', 'D_71', 'B_23', 'D_75', 'B_24', 'B_28', 'D_86', 'B_31', 'B_32', 'D_92', 'D_93', 'D_94', 'D_96', 'B_36', 'D_127']


In [24]:
# merging the target variable (y) onto the train dataset 
df_target = pd.read_csv('train_labels.csv', nrows = 20000)
df1 = pd.merge(df1, df_target, how = 'inner', on=['customer_ID'])
df1.head()

customer_ID      D_39       B_1  \
0  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.001733  0.008724   
1  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.005775  0.004923   
2  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.091505  0.021655   
3  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.002455  0.013683   
4  0000099d6bd597052cdcda90ffabf56573fe9d7c79be5f...  0.002483  0.015193   

        B_4       B_5      D_47       B_7      D_51       B_9      B_10  ...  \
0  0.080986  0.170600  0.525351  0.059416  1.335856  0.008207  0.096219  ...   
1  0.069419  0.113239  0.521311  0.057744  1.339794  0.008373  0.099804  ...   
2  0.068839  0.060492  0.524568  0.056647  1.337179  0.009355  0.134073  ...   
3  0.055630  0.166782  0.530929  0.049253  1.339909  0.006782  0.134437  ...   
4  0.038862  0.143630  0.529305  0.048918  1.341735  0.000519  0.121518  ...   

       D_86  B_31      B_32      D_92      D_93      D_94      D_96      B_36  \
0  0.007084     1  0.006626  1.006133  0.003569  0.008871  0.004950  0.009968   
1  0.006677     1  0.001854  1.005791  0.000571  0.000391  0.003180  0.003921   
2  0.001185     1  0.008686  1.005801  0.007425  0.009234  0.005433  0.001264   
3  0.003324     1  0.002478  1.007036  0.000664  0.003200  0.000063  0.002729   
4  0.001504     1  0.002199  1.002915  0.003079  0.003845  0.000535  0.009998   

      D_127  target  
0  1.003319       0  
1  1.008394       0  
2  1.009307       0  
3  1.001671       0  
4  1.009886       0  

[5 rows x 34 columns]

In [25]:
# finidng variables' correlation
X1 = df1[['D_39', 'B_1', 'B_4', 'B_5', 'D_47', 'B_7', 'D_51', 'B_9', 'B_10', 'B_11', 'B_12', 'D_58', 'B_14', 'D_60', 'D_63', 'D_65', 'B_18', 'B_21', 'D_71', 'B_23', 'D_75', 'B_24', 'B_28', 'D_86', 'B_31', 'B_32', 'D_92', 'D_93', 'D_94', 'D_96', 'B_36', 'D_127']]
corr_matrix = X1.corr()
corr_matrix.head()

D_39       B_1       B_4       B_5      D_47       B_7      D_51  \
D_39  1.000000  0.186682  0.175567  0.034771 -0.004599  0.124452  0.005021   
B_1   0.186682  1.000000  0.500963 -0.044061 -0.117911  0.626246 -0.170425   
B_4   0.175567  0.500963  1.000000  0.016175 -0.027586  0.698857 -0.171742   
B_5   0.034771 -0.044061  0.016175  1.000000  0.043975 -0.072896  0.156202   
D_47 -0.004599 -0.117911 -0.027586  0.043975  1.000000 -0.105337  0.213463   

           B_9      B_10      B_11  ...      B_28      D_86      B_31  \
D_39  0.193214 -0.016395  0.181602  ...  0.184686  0.039427 -0.086501   
B_1   0.552472 -0.034498  0.995788  ...  0.337876 -0.047622 -0.025353   
B_4   0.429494 -0.044898  0.498741  ...  0.715103 -0.021616 -0.025823   
B_5  -0.045011  0.006952 -0.038479  ...  0.287001  0.086553  0.007620   
D_47 -0.248455 -0.004065 -0.099448  ...  0.090774  0.050922  0.024117   

          B_32      D_92      D_93      D_94      D_96      B_36     D_127  
D_39  0.005429  0.031743 -0.007628 -0.014838  0.052113  0.043916  0.014094  
B_1   0.152014 -0.131280 -0.015769 -0.065572 -0.048742  0.080100 -0.143172  
B_4   0.139433 -0.124187 -0.008947 -0.071417 -0.016524  0.078066 -0.198323  
B_5  -0.021889  0.188432 -0.002322  0.031540  0.188411  0.000542  0.096221  
D_47 -0.032157  0.082343  0.077454  0.087075  0.054897 -0.016929  0.152170  

[5 rows x 31 columns]

In [26]:
upper_tri = corr_matrix.where(np.triu(np.ones(corr_matrix.shape),k=1).astype(np.bool))
print(upper_tri)

       D_39       B_1       B_4       B_5      D_47       B_7      D_51  \
D_39    NaN  0.186682  0.175567  0.034771 -0.004599  0.124452  0.005021   
B_1     NaN       NaN  0.500963 -0.044061 -0.117911  0.626246 -0.170425   
B_4     NaN       NaN       NaN  0.016175 -0.027586  0.698857 -0.171742   
B_5     NaN       NaN       NaN       NaN  0.043975 -0.072896  0.156202   
D_47    NaN       NaN       NaN       NaN       NaN -0.105337  0.213463   
B_7     NaN       NaN       NaN       NaN       NaN       NaN -0.245807   
D_51    NaN       NaN       NaN       NaN       NaN       NaN       NaN   
B_9     NaN       NaN       NaN       NaN       NaN       NaN       NaN   
B_10    NaN       NaN       NaN       NaN       NaN       NaN       NaN   
B_11    NaN       NaN       NaN       NaN       NaN       NaN       NaN   
B_12    NaN       NaN       NaN       NaN       NaN       NaN       NaN   
D_58    NaN       NaN       NaN       NaN       NaN       NaN       NaN   
B_14    NaN       NaN    

C:\Users\belli\AppData\Local\Temp\ipykernel_820\1735788728.py:1: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  upper_tri = corr_matrix.where(np.triu(np.ones(corr_matrix.shape),k=1).astype(np.bool))


In [27]:
# finding variables with a correlation greater than 72%
highest = [column for column in upper_tri.columns if any(upper_tri[column] > 0.72)]
print(highest)

['B_11', 'D_58', 'B_14', 'B_23', 'D_75']


In [28]:
# selecting 2000 rows to use the model on 
df1.sample(n = 2000)

customer_ID      D_39       B_1  \
66189   0313201ccd3fff5f5487d7c0eda3446eea85ab699fdc93...  0.001439  0.007342   
157011  075617ccf8b99b9cce35dfb652757742ae4cffd36c04f7...  0.001182  0.009015   
105870  04e5ee09af791e27ae6d294435aff8c6e0f1207f6a6340...  0.001821  0.001892   
107698  04fb0c4c6929ed49edd97ca4e150b700a0e3e5c2866742...  0.004248  0.008388   
90917   04370cccebce56b4c360b3e33e5d645b5e0a0b2ae3257b...  0.117847  0.022961   
...                                                   ...       ...       ...   
56755   02a37bca29d0d1c785266aa4a329a5e1a139d174cc4c95...  0.589228  0.067713   
41298   01e8ee4ba3e12b56acb49ee09d837b810a10453666d02a...  0.385916  0.144258   
91117   0439ce58e49842e5cfe3af935ce69141fdb992484939bd...  0.007183  0.005580   
90938   043711ec4f229f5398389a50a807edf3b2db320e0e329b...  0.001942 -0.000099   
38921   01ceb15cf91a2d895948ffebf2ceba54be8689d9b81f91...  0.009760  0.006281   

             B_4       B_5      D_47       B_7      D_51       B_9      B_10  \
66189   0.030758  0.002286  0.473886  0.030107  0.672277  0.000967  0.302729   
157011  0.116253  0.002851  0.226923  0.178630  0.338438  0.000701  0.020662   
105870  0.030102  0.008822  0.424826  0.030203  0.005280  0.006722  0.301004   
107698  0.250072  0.001181  0.500325  0.379684  0.333618  0.003076  0.014273   
90917   0.084726  0.019070  0.356155  0.178477  0.004815  0.005056  0.039207   
...          ...       ...       ...       ...       ...       ...       ...   
56755   0.031350  0.462943  0.486815  0.037127  0.336274  0.175967  0.234385   
41298   0.287132  0.031078  0.539544  0.132866  0.003816  0.561822  0.022149   
91117   0.070572  0.013099  0.970772  0.167239  0.002548  0.005649  0.025164   
90938   0.073130  0.017046  0.300943  0.040232  0.337567  0.003283  0.300839   
38921   0.029680  0.139452  0.583725  0.014326  0.668622  0.067466  0.299829   

        ...      D_86  B_31      B_32      D_92      D_93      D_94      D_96  \
66189   ...  0.007135     1  0.006044  0.008736  1.002633  0.009750  0.004188   
157011  ...  0.008754     1  0.003848  0.005037  0.009499  0.009827  0.009722   
105870  ...  0.009612     1  0.004697  0.004679  0.007498  0.003302  0.008630   
107698  ...  0.003987     1  0.002618  0.007906  0.004036  0.001693  0.001654   
90917   ...  0.006324     1  0.000575  0.002726  0.006986  0.000655  0.008620   
...     ...       ...   ...       ...       ...       ...       ...       ...   
56755   ...  0.005565     1  0.001686  0.004605  0.003939  0.007644  0.000061   
41298   ...  0.008012     1  0.007029  0.008229  0.001819  0.006146  0.009213   
91117   ...  0.007664     1  0.003138  0.008458  0.008381  0.003674  0.000022   
90938   ...  0.002566     1  0.009335  0.003371  0.002734  0.003079  0.007256   
38921   ...  0.003570     1  0.009044  1.007440  0.006469  0.008153  0.004846   

            B_36     D_127  target  
66189   0.009705  0.001618       0  
157011  0.006029  0.003785       0  
105870  0.004756  0.007723       0  
107698  0.006640  0.005553       0  
90917   0.000702  0.003058       1  
...          ...       ...     ...  
56755   0.002840  0.002185       0  
41298   0.007443  0.001729       0  
91117   0.002535  0.005125       0  
90938   0.003605  0.004636       0  
38921   0.008798  1.006231       0  

[2000 rows x 34 columns]

In [29]:
# X variables and y variable (target)
X = df1[['B_11', 'D_58', 'B_14', 'B_23', 'D_75', 'B_28']]
y = df1.target
X.shape

# split training set and testing set 
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.25)

# Running Logistic Regression model
model = sklearn.linear_model.LogisticRegression()
model.fit(X_train,y_train)
y_pred = model.predict(X_test)
print(y_pred)

[0 0 0 ... 0 0 1]


In [30]:
# Accuracy of Model
from sklearn.metrics import classification_report
matrix = classification_report(y_test, y_pred)
print(matrix)

              precision    recall  f1-score   support

           0       0.83      0.92      0.87     45144
           1       0.65      0.43      0.51     15175

    accuracy                           0.80     60319
   macro avg       0.74      0.67      0.69     60319
weighted avg       0.78      0.80      0.78     60319



# Our Process

### Why did we decided to only use (D_* ) Delinquency variables & (B_* ) Balance variables?

For the spending variables, we thought those would be unneccessary because we have the balance variables that tell us how much the customers owe after spending. We got rid of the payment variables for the same reason because balance variables will tell us the end result of what a customer needs to pay still. Risk variables could have been helpful for a prediction but delinquency variables seemed like the stronger choice for the prediction model because it tells us a history of delinquency and what could lead to a customer not filling a payment. For example, 'D_58' could have been the percentage of payments that weren't paid on time within the last 5 years. With 190 different variables and over 5.5 million customers, we needed to slim down most of the variables in order to process the program.

### How else did we clean the data?

We started by weeding out the unwanted variables (S, P, and R). Then, deleted columns with any null values because we wanted columns that had the most information to give for the prediction. Afterwards, we merged the train_labels csv file to the training dataset because we needed the 'target' variable (y) that described whether a default event would or would not occur for the predictive model. Once finding the variables with the highest correlation, we ran a logistic regression to see if a customer paid or didn't pay in which target = 1 meant they don't pay within the 120 days and target = 0 meant they do pay. 
Note: We didn't have to encode any of the categorical variables because the correlation for the variables didn't label any categorical variables as variables with the highest correlation.

# Our Conclusions

We conclude that 0.65 of the customers are probable to default payments in the future with the accuracy of 0.80. The reasons behind default payments could be financial difficulties faced by the customers also because of the overspending nature of the customers or maybe the customer’s account went through a theft or a fraud. This could be the most probable reasons of customer default payments. To find prevention from default payments American Express should set appropriate credit limits or while offering a high credit limit to the customer they should sign an agreement saying in exchange to increase the credit limit they should keep a high valuable assets from the customer. The second way can be they can sign a petition saying if the customer fails to make the payment they can produce to legal actions. In order to prevent default payments in future. Which will create a security for the bank if the customer doesn’t make the payment on time. This are all the possible ways that AMEX can use to prevent default payments from the customers.